mbf_costmap_nav/src/mbf_costmap_nav/free_pose_search.cpp

```diff
@@ -212,64 +223,78 @@ SearchSolution FreePoseSearch::findValidOrientation(const costmap_2d::Costmap2D&
 
 SearchSolution FreePoseSearch::search() const
 {
+  // restart markers
+  if (viz_)
+  {
+    viz_->deleteMarkers();
+  }
+
   const auto costmap2d = costmap_.getCostmap();
 
   // lock costmap so content doesn't change while adding cell costs
   boost::unique_lock<costmap_2d::Costmap2D::mutex_t> lock(*(costmap2d->getMutex()));
 
   std::unordered_set<int> in_queue_or_visited;
+  std::priority_queue<Cell, std::vector<Cell>, decltype(compare_strategy_)> queue(compare_strategy_);
 
   const std::vector<geometry_msgs::Point> footprint =
       safetyPadding(costmap_, config_.use_padded_fp, config_.safety_dist);
 
-  SearchSolution sol;
-  sol.pose.theta = config_.goal.theta;
-  bool outside_or_unknown{ false };
-
-  // initializing queue with the goal cell
-  Cell start;
-  if (costmap2d->worldToMap(config_.goal.x, config_.goal.y, start.x, start.y))
+  // enforce bounds if goal is outside the map; and get the pose of the center of the cell
+  Cell goal_cell;
+  geometry_msgs::Pose2D goal_cell_pose;  // this is the pose of the center of the goal cell (not the goal pose)
+  int goal_cell_x, goal_cell_y;
+  costmap2d->worldToMapEnforceBounds(config_.goal.x, config_.goal.y, goal_cell_x, goal_cell_y);
+  goal_cell.x = static_cast<unsigned int>(goal_cell_x);
+  goal_cell.y = static_cast<unsigned int>(goal_cell_y);
+  goal_cell.cost = costmap2d->getCost(goal_cell.x, goal_cell.y);
+  costmap2d->mapToWorld(goal_cell.x, goal_cell.y, goal_cell_pose.x, goal_cell_pose.y);
+
+  bool test_goal_pose{ true };
+  unsigned int dummy_x, dummy_y;
+  if (!costmap2d->worldToMap(config_.goal.x, config_.goal.y, dummy_x, dummy_y))
   {
-    start.cost = costmap2d->getCost(start.x, start.y);
+    // if goal not in bounds, we don't start the search from the goal pose
+    test_goal_pose = false;
   }
```

Comment:
Another suggestion would be to just change the queue to poses? You have to convert all the cells to poses anyway for the tolerance check, so you might as well make a single queue of poses, without special `test_goal_pose` flag?

Reply:
I don't like the second suggestion because the `queue` element shows intent and the search model. 
In our case, we are searching the free `cell` in the costmap, we update the queue according to neighboring cells, and the priority method is also an euclidean distance between cells, that could be changed to `cost` (gradient) or other custom comparison in the future. 
Therefore, changing the element, is not a small change in the code (we need to change comparison function, where cost will not be available in the future, and other parts of the code), but it also affects readability, intent and the search model. 
Besides, I don't think it is justified by the unique case of testing the goal first.